solutions/joi-15-inheritance.mdx

```diff
@@ -0,0 +1,142 @@
+---
+id: joi-15-inheritance
+source: JOI 2015
+title: Inheritance
+author: Kevin Sheng
+---
+
+At first glance, it seems Kruskal's would be more appropriate for this problem, so let's use that.
+
+Let's first give each child their own DSU data structure to store which cities are connected with their railroads,
+and then iterate through the railroads through decreasing order of price,
+giving each railroad to the first child that can accept it. This is effectively the same as going through
+each child one at a time and constructing a maximum spanning tree for each of them individually.
+
+However, linearly iterating through the entire list of DSUs would result in a
+runtime of about $\mathcal{O}(KM)$, which is too much given the bounds.
+
+To find the first child that can accept a railroad faster, let's try using binary search!
+But for binary search to actually work, we first need to prove the following:
+
+No matter what railroad we're processing,
+all children after the first child that can accept the current railroad
+can also accept said railroad, while all children *before* that child cannot.
+
```

Comment:
State this in the form of a lemma (it's unclear what you are trying to prove)

Reply:
dustin can you just rewrite this for me

Reply:
actually rereading this it's actually ok